modin/engines/base/frame/data.py

```diff
@@ -2115,14 +2116,34 @@ def groupby_reduce(
         """
         by_parts = by if by is None else by._partitions
 
-        if apply_indices is not None:
-            numeric_indices = self.axes[axis ^ 1].get_indexer_for(apply_indices)
-            apply_indices = list(
-                self._get_dict_of_block_index(axis ^ 1, numeric_indices).keys()
+        per_partition_selection = None
+        if selection is not None:
+            numeric_indices = self.axes[axis ^ 1].get_indexer_for(selection)
+            apply_indices = self._get_dict_of_block_index(axis ^ 1, numeric_indices)
+
+            sorted_selection = np.array(selection)[np.argsort(numeric_indices)]
+            internal_ids = np.cumsum(
+                (0,) + tuple((len(val) for val in apply_indices.values()))
+            )
+            per_partition_selection = OrderedDict(
+                tuple(
+                    (
+                        partition_idx,
+                        sorted_selection[internal_ids[i] : internal_ids[i + 1]],
+                    )
+                    for i, partition_idx in zip(
+                        range(len(internal_ids) - 1), apply_indices.keys()
+                    )
+                )
```

Comment:
At the Reduce phase of GroupBy, in order to handle the `as_index` parameter properly, we need to know which columns were selected at the Map phase of the corresponding partition. Being in the Reduce phase we can't compute this anyhow (`df.columns` is not the answer, because some of the selected columns could have been dropped as non-suitable for this aggregation and so won't be presented in the `df.columns`).

The way to get this info at the Reduce phase is to precompute selection for every partition and explicitly pass these selections to each partition as a function parameter.